### PR TITLE
[Bug Fix] Resolve EAGLE cuda graph IMA under PD + DP + MTP with GLM-5.1

### PR DESCRIPTION
**Pull request by zRzRzRzRzRzRzR**
Under long-running PD + DP + MTP, decode crashes with
`vectorized_gather_kernel: index out of bounds`.

Two paths feed garbage into the next cuda graph replay:

1. **Shared buffer view.** `logits_output` fields are views into
   the cuda graph's pre-allocated output buffers. Assigning them
   to `spec_info` directly means the next `replay()` overwrites
   the data `spec_info` still holds.

2. **Padded region staleness.** `EAGLEDraftCudaGraphRunner` only
   overwrites `[:raw_bs]` of its input buffers before replay.
   Padded slots `[raw_bs:bs]` retain stale values
   (e.g. `topk_index = -4886055425978319163`) and get gathered
   as OOB indices inside the graph.
   
  All guards are no-ops on healthy data.

### CODE REVIEW

**Comment by zRzRzRzRzRzRzR** [comment]
#15457 seems to mention this bug as well.

File: python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py [hunk 1]
```diff
@@ -386,8 +390,12 @@ def replay(self, forward_batch: ForwardBatch):
...
         )
         buffers.positions[:raw_num_token].copy_(forward_batch.positions)
-        buffers.topk_p[:raw_bs].copy_(forward_batch.spec_info.topk_p)
-        buffers.topk_index[:raw_bs].copy_(forward_batch.spec_info.topk_index)
+        buffers.topk_p[:raw_bs].copy_(forward_batch.spec_info.topk_p.clamp(0, 1))
+        buffers.topk_index[:raw_bs].copy_(
+            forward_batch.spec_info.topk_index.clamp(
+                0, self.model_runner.model_config.vocab_size - 1
+            )
+        )
```
Comment (JustinTong0323) [inline python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py]: The clamp silently rewrites out-of-range values to valid ones. If `topk_index` is ever genuinely OOB here, this just hides it — we'd gather token 0 instead of crashing, which is much harder to debug later.

Since the padding region is now zeroed above, the OOB root cause should already be fixed. Suggest either dropping the clamp, or replacing it with an assert (or at least a warning log when it actually changes a value).

File: python/sglang/srt/speculative/eagle_worker.py [hunk 2]
```diff
@@ -999,10 +999,10 @@ def forward_draft_extend_after_decode(self, batch: ScheduleBatch):
                 forward_batch
             )
             forward_batch.spec_info.topk_p, forward_batch.spec_info.topk_index = (
-                logits_output.topk_p,
-                logits_output.topk_index,
+                logits_output.topk_p.clone(),
+                logits_output.topk_index.clone(),
             )
-            forward_batch.spec_info.hidden_states = logits_output.hidden_states
+            forward_batch.spec_info.hidden_states = logits_output.hidden_states.clone()
```
Comment (JustinTong0323) [inline python/sglang/srt/speculative/eagle_worker.py:1161]: Worth a short comment here explaining why clone is needed, e.g. `# detach from cuda graph buffers; next replay will overwrite these slices`. Otherwise the reason gets lost and someone will "optimize" the clones away later.

**Review by JustinTong0323 — COMMENTED** [review]
Nice catch on the aliasing + stale padding. Two comments inline; also worth squashing the commits (`Update eagle_draft_cuda_graph_runner.py` etc.) before merge.

File: python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py (hunk 1 above)
Comment (zRzRzRzRzRzRzR) [reply python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py]: Okay, I will change it to assert.

File: python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py [hunk 3]
```diff
@@ -393,6 +397,17 @@ def replay(self, forward_batch: ForwardBatch):
...
+            (forward_batch.spec_info.topk_p >= 0)
+            & (forward_batch.spec_info.topk_p <= 1)
+        ).all(), "topk_p out of range [0, 1]"
+        assert (
+            (forward_batch.spec_info.topk_index >= 0)
+            & (
+                forward_batch.spec_info.topk_index
+                < self.model_runner.model_config.vocab_size
+            )
+        ).all(), "topk_index out of range [0, vocab_size)"
```
Comment (kpham-sgl) [inline python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py]: You should use `maybe_detect_oob`  to avoid cpu <-> gpu sync

File: python/sglang/srt/speculative/eagle_worker.py [hunk 4]
```diff
@@ -1155,11 +1155,12 @@ def forward_draft_extend_after_decode(self, batch: ScheduleBatch):
...
             )
+            # detach from cuda graph buffers; next replay will overwrite these slices
             forward_batch.spec_info.topk_p, forward_batch.spec_info.topk_index = (
-                logits_output.topk_p,
-                logits_output.topk_index,
+                logits_output.topk_p.clone(),
+                logits_output.topk_index.clone(),
             )
-            forward_batch.spec_info.hidden_states = logits_output.hidden_states
+            forward_batch.spec_info.hidden_states = logits_output.hidden_states.clone()
```
Comment (kpham-sgl) [inline python/sglang/srt/speculative/eagle_worker.py]: I am actually not convinced this is a bug. AFAIU for spec V1, all reader of `forward_batch.spec_info` will finish before next iter draft extend replay happen.

File: python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py [hunk 5]
```diff
@@ -384,6 +384,10 @@ def replay(self, forward_batch: ForwardBatch):
             buffers.seq_lens.fill_(self.seq_len_fill_value)
             buffers.out_cache_loc.zero_()
             buffers.positions.zero_()
+            buffers.topk_p.zero_()
+            buffers.topk_index.zero_()
+            buffers.hidden_states.zero_()
+            buffers.req_pool_indices.zero_()
```
Comment (kpham-sgl) [inline python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py:391]: I think this is correct but unsure about extra overhead

**Review by kpham-sgl — COMMENTED** [review]
Thank you for your contribution! This likely fix https://github.com/sgl-project/sglang/issues/22096 (which happen *very* rarely) as well. 

Any chances you have a reproducible script to make sure this fixes the bug.

File: python/sglang/srt/speculative/eagle_worker.py (hunk 4 above)
Comment (zRzRzRzRzRzRzR) [reply python/sglang/srt/speculative/eagle_worker.py]: Well, I have already cancelled these modifications.

File: python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py (hunk 5 above)
Comment (zRzRzRzRzRzRzR) [reply python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py:391]: Only hits the padding branch (`bs != raw_bs`), so the cost is bounded by padding size — `hidden_states` memset dominates and it's bandwidth-bound, one shot per replay. The alternative (stale OOB indices in padded slots) is what triggered the original IMA, so I'd rather eat this than make every downstream gather padding-aware.

Could tighten to a `[raw_bs:bs]` tail-only memset as a follow-up if it shows up in traces — happy to do that in a separate PR.

File: python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py [hunk 6]
```diff
@@ -384,6 +385,10 @@ def replay(self, forward_batch: ForwardBatch):
             buffers.seq_lens.fill_(self.seq_len_fill_value)
             buffers.out_cache_loc.zero_()
             buffers.positions.zero_()
+            buffers.topk_p.zero_()
+            buffers.topk_index.zero_()
+            buffers.hidden_states.zero_()
```
Comment (Qiaolin-Yu) [inline python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py:390]: The index tensor needs to be reset to 0 makes sense to me, since the padding part might cause IMA. But why do the hidden states need to be reset? My understanding is that the padding portion will be discarded anyway, so it shouldn’t have any impact at all.

Comment (zRzRzRzRzRzRzR) [reply python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py:390]: The reason I still zero  hidden_states  is about what happens inside the captured graph, not the final output:

The graph is captured at padded shape ( bs , not  raw_bs ). The captured kernels — RMSNorm, the  EagleDraftInput  projection, the attention that consumes  hidden_states  — read full  bs  rows, not a sliced view. Garbage on padded rows is fine on its own.

But if the previous replay left NaN/Inf in those padded slots (which does happen on GLM-5.1 + PD + DP + MTP — that's how I tracked this bug down), the captured RMSNorm reduces over a NaN row and contaminates scratch / shared buffers that valid lanes read in some configs. The padded output is discarded; the side effects on shared scratch inside the captured graph are not.

**Review by Qiaolin-Yu — APPROVED** [review]

**Review by hnyls2002 — APPROVED** [review]
